File: csharp/1326.minimum-number-of-taps-to-open-to-water-a-garden.cs
/*
 * @lc app=leetcode id=1326 lang=csharp
 *
 * [1326] Minimum Number of Taps to Open to Water a Garden
 */

using System;
using System.Linq;

// @lc code=start
public class Solution {
    public int MinTaps(int n, int[] ranges) {
        int ret = 0, pr = int.MinValue, r = 0;
        foreach (var curr in ranges
            .Select((int range, int index) => (Left: index - range, Right: index + range))
            .Where(t => t.Left != t.Right)
            .OrderBy(t => t.Left)) {
            if (curr.Left > r) {
                return -1;
            }
            if (curr.Left > pr) {
                pr = r;
                ++ret;
                if (curr.Right >= n) {
                    return ret;
                }
            }
            r = Math.Max(r, curr.Right);
            if (r >= n) {
                return ret;
            }
        }
        return -1;
    }
}
// @lc code=end

File: java/1326.minimum-number-of-taps-to-open-to-water-a-garden.java
/*
 * @lc app=leetcode id=1326 lang=java
 *
 * [1326] Minimum Number of Taps to Open to Water a Garden
 */

// @lc code=start
class Solution {
    public int minTaps(int n, int[] ranges) {
        int[] rights = new int[n + 1];
        int k = ranges.length;
        for (int i = 0; i < k; ++i) {
            int l = Math.max(0, i - ranges[i]), r = i + ranges[i];
            if (r == i) {
                continue;
            }
            rights[l] = Math.max(rights[l], r);
        }
        int curr = 0, r = 0, ret = 0;
        while (curr <= n && r < n) {
            ++ret;
            int nr = r;
            while (curr <= n && curr <= r) {
                nr = Math.max(nr, rights[curr]);
                ++curr;
            }
            if (nr == r) {
                return -1;
            }
            r = nr;
        }
        return ret;
    }
}
// @lc code=end

